firewood/src/merkle.rs

```diff
@@ -206,16 +205,87 @@ impl<T: TrieReader> Merkle<T> {
-    #[expect(clippy::missing_errors_doc)]
+    /// Verify that a range proof is valid for the specified key range and root hash.
+    ///
+    /// This method validates a range proof by constructing a partial trie from the proof data
+    /// and verifying that it produces the expected root hash. The proof may contain fewer
+    /// key-value pairs than requested if the peer chose to limit the response size.
+    ///
+    /// # Parameters
+    ///
+    /// * `first_key` - The requested start of the range (inclusive).
+    ///   - If `Some(key)`, verifies the proof covers keys >= this key
+    ///   - If `None`, verifies the proof starts from the beginning of the trie
+    ///
+    /// * `last_key` - The requested end of the range (inclusive).
+    ///   - If `Some(key)`, represents the upper bound that was requested
+    ///   - If `None`, indicates no upper bound was specified
+    ///   - Note: The proof may contain fewer keys than requested if the peer limited the response
+    ///
+    /// * `root_hash` - The expected root hash of the trie. The constructed partial trie
+    ///   from the proof must produce this exact hash for the proof to be valid.
+    ///
+    /// * `proof` - The range proof to verify, containing:
+    ///   - Start proof: Merkle proof for the lower boundary
+    ///   - End proof: Merkle proof for the upper boundary
+    ///   - Key-value pairs: The actual entries within the range
+    ///
+    /// # Returns
+    ///
+    /// Returns the constructed [`Merkle<Arc<ImmutableProposal>, _>`] that was built and
+    /// verified from the proof data, if the proof is valid.
+    ///
+    /// # Verification Process
+    ///
+    /// The verification follows these steps:
+    /// 1. **Structural validation**: Verify the proof structure is well-formed
+    ///    - Check that start/end proofs are consistent with the key range
+    ///    - Ensure key-value pairs are in the correct order
+    ///    - Validate that boundary proofs correctly bound the key-value pairs
+    ///
+    /// 2. **Proposal construction**: Build a proposal trie containing the proof data
+    ///    - Insert all key-value pairs from the proof
+    ///    - Incorporate nodes from the start and end proofs
+    ///    - Handle edge cases for empty ranges or partial proofs
+    ///
+    /// 3. **Hash verification**: Compute the root hash of the constructed proposal
+    ///    - The computed hash must match the provided `root_hash` exactly
+    ///    - Any mismatch indicates an invalid or tampered proof
+    ///
+    /// # Errors
+    ///
+    /// * [`api::Error::ProofError`] - The proof structure is malformed or inconsistent
+    /// * [`api::Error::InvalidRange`] - The proof boundaries don't match the requested range
+    /// * [`api::Error::IncorrectRootHash`] - The computed root hash doesn't match the expected hash
+    /// * [`api::Error`] - Other errors during proposal construction or verification
```

Comment:
nit: Not all `api::Error` types can be returned here I don't think. We currently don't do any I/O when validating a range proof, but in theory that could change